### PR TITLE
chore: Make use of splitIterator directly

### PR DESCRIPTION
**Pull request by He-Pin**
Motivation:
Use splitIterator directly

Result:
Less conversion.the current iterator is created from the splitIterator, which is another adapter, and we need double invocation hasNext and Next, both involving the adapter.

So with this directly using the split iterator, we can reduce the allocation and indirect invocation.

### CODE REVIEW

**Comment by pjfanning** [comment]
I would really prefer if there was a detailed description of why this is better. Some PRs are pretty obvious. This one isn't.

**Comment by He-Pin** [comment]
Response from my phone, the current iterator is created from the splitIterator, which is another adapter, and we need double invocation hasNext and Next, both involving the adapter.

So with this directly using the split iterator, we can reduce the allocation and indirect invocation.

**Comment by pjfanning** [comment]
> Response from my phone, the current iterator is created from the splitIterator, which is another adapter, and we need double invocation hasNext and Next, both involving the adapter.
> 
> So with this directly using the split iterator, we can reduce the allocation and indirect invocation.

Can you provide some stacktraces or code links to back this up? I checked the Javadoc for Stream iterator and spliterator and the Javadoc defines no behaviour where the iterator must call the spliterator.

**Comment by He-Pin** [comment]
@pjfanning When you call `stream.iterator`, the returns `iterator` is created from a `SplitIterator`. where it's backing by an `Adapter`
see https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Spliterators.java#L668

File: stream/src/main/scala/org/apache/pekko/stream/impl/JavaStreamSource.scala [hunk 1]
```diff
@@ -27,28 +30,32 @@ import pekko.stream.stage.{ GraphStage, GraphStageLogic, OutHandler }
...
+    new GraphStageLogic(shape) with OutHandler with Consumer[T] {
+      private var stream: java.util.stream.BaseStream[T, S] = _
+      private var splitIterator: Spliterator[T] = _
 
-      override def preStart(): Unit = {
+      final override def preStart(): Unit = {
         stream = open()
-        iter = stream.iterator()
+        splitIterator = stream.spliterator()
+        val isEmpty = splitIterator.hasCharacteristics(Spliterator.SIZED) && splitIterator.estimateSize() == 0
```
Comment (laglangyue) [inline stream/src/main/scala/org/apache/pekko/stream/impl/JavaStreamSource.scala:40]: How about
`splitIterator.estimateSize() == 0 && splitIterator.hasCharacteristics(Spliterator.SIZED)`

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/JavaStreamSource.scala:40]: Bit operation is fast

**Comment by laglangyue** [comment]
LGTM

**Comment by pjfanning** [comment]
> @pjfanning When calling `stream.iterator`, the returns `iterator` is created from a `SplitIterator`. where it's backing by an `Adapter` see https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Spliterators.java#L668

Could you provide the code where stream.iterator creates a spliterator? We may need this optimisation but we might want to only do for certain stream types. We may want to also report to OpenJDK that certain streams need to be improved so that they create iterators directly when users want iterators.

**Comment by He-Pin** [comment]
image
```java
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello World!");
        List<String> list = new ArrayList<String>();
        Iterator<String> itr = list.stream().iterator();
        System.out.println(itr.getClass());
    }
}
```

@pjfanning Just updated

**Comment by pjfanning** [comment]
@He-Pin so Java Collections class has this.

```
    default Stream<E> stream() {
        return StreamSupport.stream(spliterator(), false);
    }
```

So this explians where the spliterator comes from.

My question would still be - are we sure all the streams that we need to handle come from Collections?

**Comment by He-Pin** [comment]
We can't, but it will not be worse than the old one.

File: stream/src/main/scala/org/apache/pekko/stream/impl/JavaStreamSource.scala [hunk 2]
```diff
@@ -27,28 +30,32 @@ import pekko.stream.stage.{ GraphStage, GraphStageLogic, OutHandler }
...
+        if (isEmpty) {
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/impl/JavaStreamSource.scala:41]: do we need the isEmpty val? couldn't we just have that as the if condition?

Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/impl/JavaStreamSource.scala:41]: ok, let me inline it, but it's simpler for read.

**Review by pjfanning — APPROVED** [review]
lgtm

**Pull request merged**